Power up

JJH47E opened the pull request:
Power ups fully functional, should be easy to tweak them for different difficulties.

world.spawnBuff() & PowerUp timer needs to be checked

JJH47E committed:
Abilities Class

Added abilities class and fixed some sprite names

JJH47E committed:
Fixed detection

Uses seperate function to getNearbyObjects() as the powerups cannot be spawned in on the "object_layer" of map

JJH47E committed:
Random spawning

Buffs now spawn in at random locations, and despawn upon pickup.

Currently, buffs can spawn on top of other buffs.

In order to spawn in a buff now,  you must press "E"

JJH47E committed:
Reset resolution

JJH47E committed:
Invisibility works

Bug if 2 of the same are picked up (buff ends when first would)

JJH47E committed:
Added invincibility

JJH47E committed:
Working abilities

Invis, invincibilty, and superspeed now working at intended (other than bug in last commit)

JJH47E committed:
Abilities Finshed

All abilities working, few bugs, but nothing major

JJH47E committed:
Fixed bug

Player can now pick up as many of the same buffs as they want, and the duration will be correct

JJH47E committed:
Fixed resolution

RIP my small monitor

JJH47E committed:
Fixed bug

Can probably be done without the while loop, but it works for now

JJH47E committed:
Buffs UI

Active Buffs now apear in the UI

JJH47E committed:
Fixed UI

Tested on 1080p display

JJH47E committed:
Added docs

Provided documentation for PowerUp class

JJH47E committed:
Editted docs

markspolakovs commented on core/src/com/threecubed/auber/AuberGame.java:
```diff
@@ -11,7 +11,7 @@
   public void create() {
     atlas = new TextureAtlas("auber.atlas");
-    Gdx.graphics.setWindowedMode(1920, 1080);
+    Gdx.graphics.setWindowedMode(1980, 1080);
```
?

JJH47E replied:
My bad, I lowered it so it would fit on my screen, and typo

markspolakovs commented on core/src/com/threecubed/auber/World.java:
```diff
@@ -436,4 +443,35 @@ public void checkForEndState() {
     }
   }
+
+  public enum Abilities {
```
Any particular reason this is in `World.java` and not in its own file?

JJH47E replied:
Main reason was SystemStates & Tiles enums were not in their own files

JJH47E replied:
Has been moved in latest commit

markspolakovs commented on core/src/com/threecubed/auber/World.java:
```diff
@@ -436,4 +443,35 @@ public void checkForEndState() {
   }
+
+  public enum Abilities {
+    superspeed,
+    invisibility,
+    invincibility,
+    insta_beam,
+    vision;
```
nit: conventionally these should be UpperCamelCase

JJH47E replied:
That can be changed, I'll have to change the texture file names first

JJH47E replied:
Done

markspolakovs commented on core/src/com/threecubed/auber/World.java:
```diff
@@ -436,4 +443,35 @@ public void checkForEndState() {
+   * */
+
+  public void spawnBuff() {
+    GameEntity ent = this.getEntities().get(this.randomNumberGenerator.nextInt(this.getEntities().size()));
+    while (!(ent instanceof Civilian)) {
+      ent = this.getEntities().get(this.randomNumberGenerator.nextInt(this.getEntities().size()));
+
+    }
```
Could be rewritten as a do-while
```suggestion
    GameEntity ent;
    do {
      ent = this.getEntities().get(this.randomNumberGenerator.nextInt(this.getEntities().size()));
    } while (!(ent instanceof Civilian))
```

markspolakovs replied:
Looking at this now, it'll hardlock the game if there are no civilians - but there are no cases like that... right?

JJH47E replied:
I'll commit it for now, and add a break clause that will trigger if the number of entities in game is less that the player add the infiltrators. Then, as long as civilians don't despawn it should be fine

JJH47E replied:
[Potentially fixed in this commit](https://github.com/2020-eng1-team2/auber-pt2/commit/8874d9939413d4869ada1256ef35d0e7f6216443)

markspolakovs commented on core/src/com/threecubed/auber/entities/Player.java:
```diff
@@ -39,6 +42,14 @@
   public boolean slowed = false;
   public boolean blinded = false;
 
+  public boolean invincible = false;
+  public boolean invisible = false;
+  public boolean superspeed = false;
+  public boolean vision = false;
+  public boolean insta_beam = false;
+
```
These should probably be documented

JJH47E replied:
[Added](https://github.com/2020-eng1-team2/auber-pt2/commit/c9d5537527497b73cb366a588235802527d8a1d1)

markspolakovs commented on core/src/com/threecubed/auber/entities/PowerUp.java:
```diff
@@ -0,0 +1,133 @@
+import com.threecubed.auber.World;
+
+/**
+ * The power up entity that applies buffs to the player. Activated in @link "Player.java"
```
```suggestion
 * The power up entity that applies buffs to the player. Activated in {@link Player}
```

markspolakovs commented on core/src/com/threecubed/auber/entities/PowerUp.java:
```diff
@@ -0,0 +1,133 @@
+ * */
+public class PowerUp extends GameEntity {
+
+    private String ability;
```
Any reason this is a string, not an `Ability` enum?

JJH47E replied:
It gets the texture from the atlas using the name which needs to be a string, easy enough to change though

JJH47E replied:
Changed, ability now takes type Abilities rather than String

markspolakovs reviewed (changes requested):
Something something write tests :p (although I know it's not exactly easy)

JJH47E committed:
Update core/src/com/threecubed/auber/entities/PowerUp.java

Co-authored-by: Marks Polakovs <github@markspolakovs.me>

JJH47E committed:
Begun removing String use

Begun removing String use in PowerUp

JJH47E committed:
Update core/src/com/threecubed/auber/World.java

Co-authored-by: Marks Polakovs <github@markspolakovs.me>

JJH47E committed:
Changed file names

JJH47E committed:
Merge branch 'power-up' of https://github.com/2020-eng1-team2/auber-pt2 into power-up

JJH47E committed:
Temporary Bug fix

Make a test for this?

JJH47E committed:
Documentation

Added documentation

JJH47E committed:
Abilities moved to seperate file

Abilities enum ow in seperate file, PowerUp class now uses the enum value rather than string

JJH47E committed:
Added some tests

PowerUpTest does not work currently

JJH47E committed:
Renamed test functions

Still broken!!

JJH47E committed:
Delete PowerUpTest.java
